### PR TITLE
*: Setup Global Resource Controller

### PR DESCRIPTION
**Pull request by nolouch**
Signed-off-by: nolouch <nolouch@gmail.com>

### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/40731

Problem Summary:

### What is changed and how it works?

``` commit
- Introduce the Global Resource Controller to TiDB, and resource controller is power by a Distributed Token Bucket.
- Global Resource Controller is used to control the whole cluster level resource limit. it can use the abstruct request unit as the limit unit.
```
### Check List

wait PR: https://github.com/tikv/client-go/pull/648
Tests 

- [x] Unit test
- [x] Manual test (add detailed scripts or steps below)
  record in https://asciinema.org/a/552982

Side effects

- [x] Performance regression: May reduce a little performance
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Comment by nolouch** [comment]
PTAL @JmPotato @BornChanger @hawkingrei

File: sessionctx/variable/tidb_vars.go [hunk 1]
```diff
@@ -1265,6 +1265,16 @@ var (
...
+	SetGlobalResourceControl atomic.Pointer[func(bool)]
+)
+
+// Hooks functions for Cluster Resource Control.
+var (
+	// SetupGlobalResourceControlFunc is the function registered by tikv_driver to set cluster resource control.
+	EnableGlobalResourceControlFunc func() = func() {}
+	// UnsetGlobalResourceControlFunc is the function registered by tikv_driver to unset cluster resource control.
+	DisableGlobalResourceControlFunc func() = func() {}
 )
```
Comment (HuSharp) [inline sessionctx/variable/tidb_vars.go:1278]: Should be consistent with the comments?

**Comment by hawkingrei** [comment]
@nolouch  Please fix it.

```
store/driver/tikv_driver.go:98:6: ineffectual assignment to ok (ineffassign)
```

**Comment by nolouch** [comment]
PTAL again @hawkingrei  thanks!

File: sessionctx/variable/sysvar_test.go [hunk 2]
```diff
@@ -1123,18 +1136,22 @@ func TestTiDBEnableResourceControl(t *testing.T) {
 
 	// Default true
```
Comment (BornChanger) [inline sessionctx/variable/sysvar_test.go:1137]:
```suggestion
	// Default false
```

**Comment by hawkingrei** [comment]
We can sync our branch to the latest master every time when we push it. so that we can use the latest cache to accelerate our test. the old cache may be deleted.

**Review by BornChanger — APPROVED** [review]
LGTM~

File: sessionctx/variable/tidb_vars.go [hunk 3]
```diff
@@ -1265,6 +1265,16 @@ var (
 	SetExternalTimestamp func(ctx context.Context, ts uint64) error
 	// GetExternalTimestamp is the func registered by staleread to get externaltimestamp from pd
 	GetExternalTimestamp func(ctx context.Context) (uint64, error)
+	//SetGlobalResourceControl is the func registered by domain to set cluster resource control.
```
Comment (HuSharp) [inline sessionctx/variable/tidb_vars.go:1268]: Seems like lack of a blank

Comment (nolouch) [reply sessionctx/variable/tidb_vars.go]: It's strange no lint report. addressed.

**Review by hawkingrei — APPROVED** [review]

**Review by cfzjywxk — APPROVED** [review]

**Comment by hawkingrei** [comment]
```
[2023/01/20 23:23:44.747 +08:00] [INFO] [domain_sysvars.go:79] ["set resource control"] [enable=false]

[2023/01/20 23:24:14.749 +08:00] [INFO] [domain_sysvars.go:79] ["set resource control"] [enable=false]

[2023/01/20 23:24:44.751 +08:00] [INFO] [domain_sysvars.go:79] ["set resource control"] [enable=false]

[2023/01/20 23:25:14.754 +08:00] [INFO] [domain_sysvars.go:79] ["set resource control"] [enable=false]

[2023/01/20 23:25:44.757 +08:00] [INFO] [domain_sysvars.go:79] ["set resource control"] [enable=false]

[2023/01/20 23:26:14.760 +08:00] [INFO] [domain_sysvars.go:79] ["set resource control"] [enable=false]

[2023/01/20 23:26:44.762 +08:00] [INFO] [domain_sysvars.go:79] ["set resource control"] [enable=false]

[2023/01/20 23:27:14.765 +08:00] [INFO] [domain_sysvars.go:79] ["set resource control"] [enable=false]

[2023/01/20 23:27:44.768 +08:00] [INFO] [domain_sysvars.go:79] ["set resource control"] [enable=false]

[2023/01/20 23:28:14.770 +08:00] [INFO] [domain_sysvars.go:79] ["set resource control"] [enable=false]

[2023/01/20 23:28:44.773 +08:00] [INFO] [domain_sysvars.go:79] ["set resource control"] [enable=false]

[2023/01/20 23:29:14.776 +08:00] [INFO] [domain_sysvars.go:79] ["set resource control"] [enable=false]

[2023/01/20 23:29:44.779 +08:00] [INFO] [domain_sysvars.go:79] ["set resource control"] [enable=false]

[2023/01/20 23:30:14.782 +08:00] [INFO] [domain_sysvars.go:79] ["set resource control"] [enable=false]

[2023/01/20 23:30:44.785 +08:00] [INFO] [domain_sysvars.go:79] ["set resource control"] [enable=false]

[2023/01/20 23:31:14.787 +08:00] [INFO] [domain_sysvars.go:79] ["set resource control"] [enable=false]

[2023/01/20 23:31:44.790 +08:00] [INFO] [domain_sysvars.go:79] ["set resource control"] [enable=false]

[2023/01/20 23:32:14.793 +08:00] [INFO] [domain_sysvars.go:79] ["set resource control"] [enable=false]

[2023/01/20 23:32:44.797 +08:00] [INFO] [domain_sysvars.go:79] ["set resource control"] [enable=false]
```

I find many log about resource control. Has it a problem?

**Pull request merged**